elastic/kibana
# [Index Management] Support managed_by field and fix never delete data logic

Closes https://github.com/elastic/kibana/issues/167414
Closes https://github.com/elastic/kibana/issues/167416
Closes https://github.com/elastic/kibana/issues/167805

## Summary
This PR introduces a couple of separate states for the edit data retention modal, that can be triggered depending on the state of the datastream.

##### How to test
1. Start elasticsearch with `yarn es snapshot --license=trial -E path.repo=/tmp/snapshots` and kibana with `yarn start`
2. Navigate to console and execute the snippet for the case you wanna test
3. Navigate to index management and click on data streams tab. 
4. Verify that:
  * When the datastream is fully managed by ILM, edit option shouldnt be shown in the manage options and a little warning callout should be shown in the details flyout
  * When is fully managed by DSL, no callouts should be shown
  * When is managed by DSL, but has ILM backing indices we allow editing but show a callout saying that some indices are only managed by ILM

<details>
<summary>Create a datastream that is fully managed by ILM</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}
```
</details>

<details>
<summary>Create a datastream that is unmanaged</summary>

```
PUT _index_template/my-data-stream-template
{
    "index_patterns": [ "my-data-stream*" ],
    "data_stream": {},
    "priority": 400,
    "template" : {
        "lifecycle": { 
           "data_retention": "7d",
           "enabled": false 
        }
    }
}


POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SYKENET-000001",
  "location": "swamp",
  "temperature": 32.4,
  "humidity": 88.9
}
```
</details>

<details>
<summary>Create a datastream that managed by DSL but has ILM indices</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}



PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm",
      "index.lifecycle.prefer_ilm": false
    },
    "lifecycle": {
      "data_retention": "7d"
    }
  }
}

POST my-data-stream2/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}

PUT _data_stream/my-data-stream/_lifecycle
{
  "data_retention": "7d"
}
```
</details>

<details>
<summary>Create a datastream fully managed by DSL</summary>

```
PUT _cluster/settings
{
  "persistent": {
    "indices.lifecycle.poll_interval": "1s",
    "data_streams.lifecycle.poll_interval": "1s",
    "cluster.lifecycle.default.rollover": "min_docs=2,max_docs=2"
  }
}

PUT _ilm/policy/rollover-w-ilm
{
  "policy": {
    "phases": {
      "hot": {
        "actions": {
          "rollover": {
            "max_docs": "1"
          }
        }
      }
    }
  }
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm"
    }
  }
}

POST my-data-stream/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}

PUT _index_template/my-data-stream-template
{
  "index_patterns": ["my-data-stream*"],
  "data_stream": {},
  "priority": "400",
  "template": {
    "settings": {
      "index.lifecycle.name": "rollover-w-ilm",
      "index.lifecycle.prefer_ilm": false
    },
    "lifecycle": {
      "data_retention": "7d"
    }
  }
}

POST my-data-stream2/_doc?refresh
{
  "@timestamp": "2023-05-06T16:21:15.000Z",
  "sensor_id": "SKYNET-000001",
  "location": "swamp"
}
```
</details>

<details>
<summary>Screenshots</summary>

// Fully managed by ILM
![Screenshot 2023-10-25 at 10 15 52](https://github.com/elastic/kibana/assets/1191206/f843eae4-4aa3-416b-a3c2-b68c8e2831f9)
// --> also when its fully managed by ILM the computed lifecycle is managed by ILM, so it will be strikedthrough in the table and flyout
![Screenshot 2023-10-26 at 17 25 12](https://github.com/elastic/kibana/assets/1191206/cf2a01da-5ecc-4d5e-8cea-d327ee4b3321)


// Fully managed by DSL
![Screenshot 2023-10-25 at 10 20 45](https://github.com/elastic/kibana/assets/1191206/900b6d38-7d62-4b5d-9f4f-a71b5acda3b8)
![Screenshot 2023-10-26 at 10 17 36](https://github.com/elastic/kibana/assets/1191206/03e261ed-a6ff-4a02-b724-75c075dd9528)
![Screenshot 2023-10-26 at 10 18 18](https://github.com/elastic/kibana/assets/1191206/684cbfd8-b35c-4fa6-a5db-64eecd4f73b6)



// Managed by DSL but has backing indices managed by ILM
![Screenshot 2023-10-26 at 10 16 29](https://github.com/elastic/kibana/assets/1191206/8d504dbd-2391-470b-9a19-d5610d128018)
![Screenshot 2023-10-26 at 10 16 54](https://github.com/elastic/kibana/assets/1191206/828984ee-f71e-4ee9-a423-963331b029b4)


</details>
